test/e2e/pkg/grammar/checker.go

```diff
@@ -161,33 +154,33 @@ func (g *Checker) verifyCleanStart(execution string) []*Error {
 		expectedTokens := strings.Join(exp, ",")
 		unexpectedToken := err.Token.TypeID()
 		e := &Error{
-			description: fmt.Sprintf("Invalid clean-start execution: parser was expecting one of [%v], got [%v] instead.", expectedTokens, unexpectedToken),
+			description: fmt.Sprintf("Invalid execution: parser was expecting one of [%v], got [%v] instead.", expectedTokens, unexpectedToken),
 			height:      err.Line - 1,
 		}
 		errors = append(errors, e)
 	}
-	return errors
-}
-
-// verifyRecovery verifies if a specific execution is a valid recovery execution.
-func (g *Checker) verifyRecovery(execution string) []*Error {
-	errors := make([]*Error, 0)
-	lexer := recovery_lexer.New([]rune(execution))
-	_, errs := recovery_parser.Parse(lexer)
-	for _, err := range errs {
-		exp := []string{}
-		for _, ex := range err.Expected {
-			exp = append(exp, ex)
-		}
-		expectedTokens := strings.Join(exp, ",")
-		unexpectedToken := err.Token.TypeID()
-		e := &Error{
-			description: fmt.Sprintf("Invalid recovery execution: parser was expecting one of [%v], got [%v] instead.", expectedTokens, unexpectedToken),
-			height:      err.Line - 1,
+	if len(errors) != 0 {
+		return errors
+	}
+	eType := symbols.NT_Recovery
+	if isCleanStart {
+		eType = symbols.NT_CleanStart
+	}
+	roots := bsrForest.GetRoots()
+	for _, r := range roots {
```

Comment:
Question: is this returning _all_ possible roots and all possible way the grammar could be applied to the trace?

This is important because of the following. I'll explain it with an example:

* Imagine that the last recovery trace (so an _actual_ recovery) contains the following trace: `InitChain`, `FinalizeBlock, `Commit`.
* This trace, obtained from a recovering process, can be misinterpreted as a clean-start trace

Hence it is important when we are checking for the root(s) [I understand: the starting non-terminal(s)]

Reply:
My understanding is that the answer to your question is `Yes`. This is what is written in the `GetRoots` description: `// GetRoots returns all the roots of parse trees of the start symbol of the grammar`. Moreover, in my unit tests I am actually checking whether the execution you just described is passing the check as both the clean-start and recovery execution. See tests `empty-block-1` and `empty-block-2` in the `TestVerify` function in `pkg/grammar/checker_test.go` file.